app/controllers/patients_controller.rb

```diff
@@ -542,6 +548,105 @@ def update_monitoring_fields(patient, params, household, propagation)
     patient.monitoring_history_edit(history_data, diff_state)
   end
 
+  # Update the patient's follow-up flag fields
+  def update_follow_up_flag
+    redirect_to(root_url) && return unless current_user.can_edit_patient? && !current_user.role?(Roles::ENROLLER)
+
+    patient = current_user.get_patient(params.permit(:id)[:id])
+    redirect_to(root_url) && return if patient.nil?
+
+    update_follow_up_flag_fields(patient, params)
+  end
+
+  # Make updates to follow-up flag reason and/or note and create corresponding History items.
+  #
+  # patient - The Patient to update.
+  # params - The request params.
+  def update_follow_up_flag_fields(patient, params)
+    clear_flag = params.permit(:clear_flag)[:clear_flag]
+    history_data = {}
+    if clear_flag
+      clear_flag_reason = params.permit(:clear_flag_reason)[:clear_flag_reason]
+      clear_follow_up_flag(patient, clear_flag_reason)
+    else
+      follow_up_reason_before = patient.follow_up_reason
+      follow_up_note_before = patient.follow_up_note
+      follow_up_reason = params.permit(:follow_up_reason)[:follow_up_reason]
+      follow_up_note = params.permit(:follow_up_note)[:follow_up_note]
+
+      # Apply and save updates to the db
+      patient.update!(follow_up_reason: follow_up_reason, follow_up_note: follow_up_note)
+
+      # Handle creating history items based on the updates
+      history_data = {
+        created_by: current_user.email,
+        patient: patient,
+        follow_up_reason: follow_up_reason,
+        follow_up_note: follow_up_note,
+        follow_up_reason_before: follow_up_reason_before,
+        follow_up_note_before: follow_up_note_before
+      }
+
+      History.follow_up_flag_edit(history_data)
+    end
+
+    # If not applying to household, return
+    apply_to_household_ids = params.permit(apply_to_household_ids: [])[:apply_to_household_ids]
```

Comment:
I think we have to deal with the same concurrency problem for bulk updates, which don't use this logic for dependent updates but are individually passed to this method as the patient. maybe a potential solution could be to change the way this method is called for dependents in line 454-456 so that they are passed in the `apply_to_household_ids` param

Reply:
I'm wondering if we can consolidate this race condition checking logic into some separate method because it's used in multiple places and will probably be used for future features as well

Reply:
@timwongj So good news is that it doesn't appear we'll have the same concurrency issue that exists in the `ApplyToHousehold` component for the bulk action "Apply this change to the entire household that these monitorees are responsible for" toggle. 

For the `ApplyToHousehold` component, the user is able to both view and select the specific household members to apply the change to, which results in the concurrency issue previous addressed. For the bulk action Apply to Household toggle, it's simply a toggle that doesn't allow the user to view or specify which particular household members to apply the change to. The application doesn't identify and select the applicable houshold members until it reaches this part of the code. So a user could remove a household member for a HOH that a bulk action is about to be done on and that removed household member will not be impacted by the bulk action as we would hope. I just tested this out for a couple different bulk actions including the follow-up flag and this seems like it won't be an issue.

This being said, I will like your suggestion of breaking this part of the code out into a reusable function. I will look to complete that with my next commit.

Reply:
ah, you are right. thanks for refactoring the code as well, awesome work!